clap_generate/src/shell.rs

```diff
@@ -19,36 +21,160 @@ pub enum Shell {
 
     fn from_str(s: &str) -> Result<Self, Self::Err> {
-        match s.to_ascii_lowercase().as_str() {
-            "bash" => Ok(Shell::Bash),
-            "elvish" => Ok(Shell::Elvish),
-            "fish" => Ok(Shell::Fish),
-            "powershell" => Ok(Shell::PowerShell),
-            "zsh" => Ok(Shell::Zsh),
-            _ => Err(String::from(
-                "[valid values: bash, elvish, fish, powershell, zsh]",
-            )),
+        MAPPINGS
+            .iter()
+            .filter_map(|(current_str, current_value)| {
+                current_str.eq_ignore_ascii_case(s).then(|| *current_value)
+            })
+            .next()
+            .ok_or_else(|| String::from(VALID_VALUES))
+    }
+}
+
+// Hand-rolled so it can work even when `derive` feature is disabled
+impl ArgEnum for Shell {
+    const VARIANTS: &'static [&'static str] = &VARIANTS;
+
+    /// Parse an argument into `Self`.
+    fn from_str(input: &str, case_insensitive: bool) -> Result<Self, String> {
+        let eq = if case_insensitive {
+            str::eq_ignore_ascii_case
+        } else {
+            PartialEq::eq
+        };
+        MAPPINGS
+            .iter()
+            .filter_map(|(current_str, current_value)| {
+                eq(current_str, input).then(|| *current_value)
+            })
+            .next()
+            .ok_or_else(|| String::from(VALID_VALUES))
+    }
+
+    fn as_arg(&self) -> Option<&'static str> {
+        MAPPINGS
+            .iter()
+            .filter_map(|(current_str, current_value)| {
+                (self == current_value).then(|| *current_str)
+            })
+            .next()
+    }
+}
+
+impl crate::Generator for Shell {
+    fn file_name(&self, name: &str) -> String {
+        match self {
+            Shell::Bash => crate::generators::Bash.file_name(name),
+            Shell::Elvish => crate::generators::Elvish.file_name(name),
+            Shell::Fish => crate::generators::Fish.file_name(name),
+            Shell::PowerShell => crate::generators::PowerShell.file_name(name),
+            Shell::Zsh => crate::generators::Zsh.file_name(name),
+        }
+    }
+
+    fn generate(&self, app: &clap::App, buf: &mut dyn std::io::Write) {
+        match self {
+            Shell::Bash => crate::generators::Bash.generate(app, buf),
+            Shell::Elvish => crate::generators::Elvish.generate(app, buf),
+            Shell::Fish => crate::generators::Fish.generate(app, buf),
+            Shell::PowerShell => crate::generators::PowerShell.generate(app, buf),
+            Shell::Zsh => crate::generators::Zsh.generate(app, buf),
+        }
+    }
+}
+
+const VALID_VALUES: &str = "[valid values: bash, elvish, fish, powershell, zsh]";
+
+const VARIANTS: [&str; 5] = ["bash", "elvish", "fish", "powershell", "zsh"];
+
+const MAPPINGS: [(&str, Shell); 5] = [
+    ("bash", Shell::Bash),
+    ("elvish", Shell::Elvish),
+    ("fish", Shell::Fish),
+    ("powershell", Shell::PowerShell),
+    ("zsh", Shell::Zsh),
+];
```

Comment:
Let's move these statics to top the file after imports.

Reply:
If these were `pub`, I'd agree.  With these being implementation details, I think its better to be at the end so that the "top level" item is at the top of the file, making it easier for people reading the code to get a "jumping off point" for understanding this module.